x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskTests.java

```diff
@@ -5,205 +5,613 @@
+
+    public void testCoordinateReads() {
+        ShardFollowNodeTask task = createShardFollowTask(8, 8, 8, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(8));
+        assertThat(shardChangesRequests, contains(new long[][]{
+            {0L, 8L}, {8L, 8L}, {16L, 8L}, {24L, 8L}, {32L, 8L}, {40L, 8L}, {48L, 8L}, {56L, 8L}}
+        ));
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(8));
+        assertThat(status.getLastRequestedSeqno(), equalTo(63L));
+    }
+
+    public void testWriteBuffer() {
+        // Need to set concurrentWrites to 0, other the write buffer gets flushed immediately:
+        ShardFollowNodeTask task = createShardFollowTask(64, 1, 0, 32, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(1));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(64L));
+
+        shardChangesRequests.clear();
+        // Also invokes the coordinatesReads() method:
+        task.innerHandleReadResponse(0L, 63L, generateShardChangesResponse(0, 63, 0L, 128L));
+        assertThat(shardChangesRequests.size(), equalTo(0)); // no more reads, because write buffer is full
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(0));
+        assertThat(status.getNumberOfConcurrentWrites(), equalTo(0));
+        assertThat(status.getLastRequestedSeqno(), equalTo(63L));
+        assertThat(status.getLeaderGlobalCheckpoint(), equalTo(128L));
+    }
+
+    public void testMaxConcurrentReads() {
+        ShardFollowNodeTask task = createShardFollowTask(8, 1, 1, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(1));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(8L));
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(1));
+        assertThat(status.getLastRequestedSeqno(), equalTo(7L));
+    }
+
+    public void testTaskCancelled() {
+        ShardFollowNodeTask task = createShardFollowTask(64, 1, 1, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(1));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(64L));
+
+        shardChangesRequests.clear();
+        synchronized (task) {
+            task.updateLeaderGlobalCheckpoint(128L);
```

Comment:
We can setup other scenarios to test the cancellation if we remove `updateLeaderGlobalCheckpoint`. For example, make the read limits reached, then cancel, then verify that we won't issue any read request.